### PR TITLE
More adapters added: deepseek,  llama, nova, opeai, qwen. 

### PR DESCRIPTION
**Pull request by Michael-Hollins**
Also fixed a bug for mistral such that mistral models using regex now point to the function_calling module rather than the old mistral adapter.

Tests still required.

### CODE REVIEW

**Comment by Michael-Hollins** [comment]
@gillespied to review

**Comment by gillespied** [comment]
@Michael-Hollins adding the extra models will make the library a lot more useful. I've got a few requested changes to bring things in line with the repo's existing patterns before we merge:

Must-fix:

1. Add docstrings to all public methods - the repo uses Google-style docstrings (see AGENTS.md and anthropic.py for reference). The refactored mistral_function_calling.py has had its docstrings removed, and the new adapters are missing them on most methods.

2. Add unit tests - CI requires passing pytest with coverage. Each existing adapter has a corresponding test file. Please add tests for the new adapters (at minimum: build_tool, prepare_model_input, and validate_result for each, covering both success and failure paths).

3. Deduplicate the OpenAI-compatible adapters - deepseek.py, openai_oss.py, and qwen.py are near-identical. Consider a shared base class (e.g., OpenAICompatibleAdapter) that the three inherit from with only the differences overridden. Similarly, the $defs inlining logic is duplicated across three files - extract it to a utility or base class method.

4. Don't use setattr to dynamically add fields - llama.py and nova.py do setattr(model_input, "prompt", ...) which bypasses typing and will fail mypy. If ModelInput needs new fields for these providers, let's add them to the dataclass (or discuss an alternative approach).

5. Preserve the existing validation checks in mistral_function_calling.py - the original checked finish_reason, tool count, and tool name match. Removing these makes debugging batch failures much harder. Please keep the granular checks with their debug log messages.

6. Don't remove MistralAdapter from the top-level exports - this is a breaking change for existing consumers. Keep it exported (even if you also export MistralFunctionAdapter).

Minor/style: you need to run these through ruff for linting.

**Comment by Michael-Hollins** [comment]
Thanks @gillespied for that feedback. I've actioned all your points. 

A possible next step on a new branch would be improving the result_validation methods in the new adapters compared against some real batch job return objects. 

Also, running:

uv run mypy src/

Led to some errors with the batch_inferer.py and structured_batch_inferer.py, but as these aren't mine I haven't touched them. 

Please let me know if anything further is needed.

File: src/llmbo/models.py [hunk 1]
```diff
@@ -36,29 +36,45 @@ class ModelInput:
...
     # These are required
-    messages: list[dict]
-    anthropic_version: str = "bedrock-2023-05-31"
-    max_tokens: int = 2000
+    messages: list[dict] | None = None
+    anthropic_version: str | None = "bedrock-2023-05-31"
+    max_tokens: int | None = 2000
 
-    system: str | None = None
+    system: str | list[dict[str, Any]] | None = None
```
Comment (gillespied) [inline src/llmbo/models.py:81]: This is a bit of a lie now that messages isnt "required". In fact a lot of the attributes are now optional and so we are losign the point of having one. Its probably worth a refactor to split this out into multiple classes and have some logic to select the correct one. 

I dont think its required for this PR, but please open an issue.

**Review by gillespied — CHANGES_REQUESTED** [review]
Great job on this. A couple more things:

- update the readme to mention your new adapters. 
- open an issue for my mypy bits
- open an issue to refactor model input. 

Then happy to merge.

File: src/llmbo/models.py (hunk 1 above)
Comment (gillespied) [reply src/llmbo/models.py:81]: and add yourself to the authors in pyproject.

**Review by gillespied — APPROVED** [review]

**Comment by Michael-Hollins** [comment]
Hi @gillespied . I came across a bug with Nova 2 which uses the Converse API. I believe it's all fixed so I've merged it into this PR, but please take a look. Then, maybe we merge into main and close the PR?
